### PR TITLE
Hact dashboard changes

### PR DESCRIPTION
**Pull request by ntrncic**

### CODE REVIEW

File: EquiTrack/partners/models.py [hunk 1]
```diff
@@ -239,95 +272,91 @@ def hact_min_requirements(self):
...
-        if self.partner_type == u'Government':
+        total = 0
+        if partner.partner_type == u'Government':
             total = GovernmentInterventionResult.objects.filter(
-                intervention__partner=self,
+                intervention__partner=partner,
                 year=year).aggregate(
                 models.Sum('planned_amount')
             )['planned_amount__sum'] or 0
+            if budget_record:
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: concerned of double adding on budget mods

File: EquiTrack/partners/models.py [hunk 2]
```diff
@@ -239,95 +272,91 @@ def hact_min_requirements(self):
...
-
-        pv = self.cp_cycle_trip_links.filter(
-                trip__travel_type=Trip.PROGRAMME_MONITORING
-            ).exclude(
-                trip__status__in=[Trip.CANCELLED, Trip.COMPLETED]
-            ).count() or 0
-
-
-        return pv
+            if budget_record:
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: same here

File: EquiTrack/partners/models.py [hunk 3]
```diff
@@ -359,38 +388,62 @@ def trips(self):
...
         )
 
-    @property
-    def programmatic_visits(self):
+    @classmethod
+    def planned_visits(cls, partner, trip=None):
+        from trips.models import Trip
+        # planned visits
+        pv = 0
+        pv = partner.cp_cycle_trip_links.filter(
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: here let's actually just get the planned trips from the interventions

File: EquiTrack/partners/models.py [hunk 4]
```diff
@@ -359,38 +388,62 @@ def trips(self):
...
 
-    @property
-    def follow_up_flags(self):
+        if trip and trip.travel_type == Trip.SPOT_CHECK \
+                and trip.status in [Trip.COMPLETED]:
+            sc += 1
+        partner.update_hact_value('spot_checks', sc)
+
+    @classmethod
+    def follow_up_flags(cls, partner, tr=None):
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: I think this is for action points, let's revisit this

File: EquiTrack/partners/models.py [hunk 5]
```diff
@@ -1323,6 +1407,14 @@ def save(self, **kwargs):
 
         super(PartnershipBudget, self).save(**kwargs)
 
+        if self.unicef_cash:
+            if self.pk:
```
Comment (robertavram) [inline EquiTrack/partners/models.py]: we need to make this a transactional save

File: EquiTrack/partners/tests/test_models.py [hunk 6]
```diff
@@ -147,14 +147,5 @@ def setUp(self):
...
-
-    # The total cash transferred and actual cash transferred for the partner now comes from vision
-    # def test_actual_cash_transferred(self):
-    #     total = self.intervention.partner.actual_cash_transferred
-    #     self.assertEqual(total, 40000)
-
-    # def test_total_cash_transferred(self):
-    #     total = self.intervention.partner.total_cash_transferred
-    #     self.assertEqual(total, 80000)
+        self.intervention.partner.planned_cash_transfers(self.intervention.partner)
```
Comment (robertavram) [inline EquiTrack/partners/tests/test_models.py]: we will need to update this test as well

File: EquiTrack/partners/models.py [hunk 7]
```diff
@@ -180,6 +181,15 @@ def __unicode__(self):
     def latest_assessment(self, type):
         return self.assessments.filter(type=type).order_by('completed_date').last()
 
+    def update_hact_value(self, key, value):
```
Comment (ntrncic) [inline EquiTrack/partners/models.py]: done!

File: EquiTrack/partners/models.py [hunk 8]
```diff
@@ -202,24 +212,47 @@ def micro_assessment_needed(self):
...
-            and self.type_of_assessment in ['Micro Assessment', 'Negative Audit Results'] \
+        micro_assessment = partner.assessments.filter(type=u'Micro Assessment').order_by('completed_date').last()
+        if assessment:
+            if assessment.completed_date > micro_assessment.completed_date:
+                micro_assessment = assessment
+        if partner.type_of_assessment == 'High Risk Assumed':
+            partner.update_hact_value('micro_assessment_needed', 'Yes')
+        elif partner.planned_cash_transfers > 100000.00 \
+            and partner.type_of_assessment == 'Simplified Checklist' or partner.rating == 'Not Required':
+            partner.update_hact_value('micro_assessment_needed', 'Yes')
```
Comment (ntrncic) [inline EquiTrack/partners/models.py]: done!

File: EquiTrack/partners/models.py [hunk 9]
```diff
@@ -519,6 +572,27 @@ def __unicode__(self):
...
         )
 
+    @transaction.atomic
+    def save(self, **kwargs):
+        # set partner last micro assessment
+        if self.type == u'Micro Assessment' and self.completed_date:
+            if self.pk:
+                prev_assessment = Assessment.objects.get(id=self.id)
+                if prev_assessment.completed_date and prev_assessment.completed_date != self.completed_date:
+                    self.partner.micro_assessment_needed(self.partner)
```
Comment (ntrncic) [inline EquiTrack/partners/models.py]: done

File: EquiTrack/partners/models.py [hunk 10]
```diff
@@ -519,6 +572,27 @@ def __unicode__(self):
...
+            else:
+                self.partner.micro_assessment_needed(self.partner)
+
+        elif self.type == u'Scheduled Audit report' and self.completed_date:
+            if self.pk:
+                prev_assessment = Assessment.objects.get(id=self.id)
+                if prev_assessment.completed_date and prev_assessment.completed_date != self.completed_date:
+                    self.partner.audit_needed(self.partner)
```
Comment (ntrncic) [inline EquiTrack/partners/models.py]: done!

**Pull request merged**